### ISSUE TITLE
Safari loads wasm twice

### ISSUE THREAD

**Issue by 617a7aa** [leptos-rs/leptos#1450]
**Describe the bug**
In Safari (Version 17.0 (19616.1.20.11.3)), Leptos loads the wasm file twice. [Production example](https://valera.co) - source code [here](https://github.com/valeralabs/web/tree/mk3) (mk3 branch)

**Leptos Dependencies**

Please copy and paste the Leptos dependencies and features from your `Cargo.toml`.

For example:
```toml
leptos = { git = "https://github.com/leptos-rs/leptos", default-features = false, features = [
	"rkyv",
] }
leptos_meta = { git = "https://github.com/leptos-rs/leptos", default-features = false }
leptos_actix = { git = "https://github.com/leptos-rs/leptos", default-features = false, optional = true }
leptos_router = { git = "https://github.com/leptos-rs/leptos", default-features = false }
```

**To Reproduce**
1. Go to ['valera.co'](https://valera.co) in Safari
2. Open web console, go to networking
3. Reload page
4. Wasm gets loaded once from XHR in HTML, twice from web.js from fetch

**Expected behavior**
One load of wasm

**Screenshots**

<img width="100%" alt="Screenshot 2023-07-27 at 11 44 46" src="https://github.com/leptos-rs/leptos/assets/67651251/807907dd-a8eb-4722-9649-680c1ace7f5d">

**Comment by gbj** [comment]
Currently Leptos inserts the following to begin prefetching the JS and WASM files, then loading the JS to bootstrap the WASM module and hydrating the page:
```html
<link rel="modulepreload" href="/pkg/web.js">
<link rel="preload" href="/pkg/web.wasm" as="fetch" type="application/wasm" crossorigin="">
<script type="module">
  import init, { hydrate } from '/pkg/web.js'; init('/pkg/web.wasm').then(hydrate);
</script>
```
This works correctly in Firefox and Chrome: each sends only a single request to begin preloading the module, then uses that same request to instantiate streaming in `init('/pkg/web.wasm')`.

I don't know why Safari behaves differently.

[Per MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/preload), `as="fetch"` is correct for WASM:
> `fetch`: Resource to be accessed by a fetch or XHR request, such as an ArrayBuffer, WebAssembly binary, or JSON file.

If you or anyone else is aware of a way to make Safari behave correctly, or any explanation for why it behaves differently from the other browsers, I'd be happy to hear it. But as far as I know the HTML we're sending is correct.

**Comment by 617a7a** [comment]
https://stackoverflow.com/a/63814972

The different CORS modes are messing up the request caching.

**Comment by gbj** [comment]
Do you mean the `crossorigin`? Removing it does not prevent Safari from fetching twice.

**Comment by 617a7a** [comment]
Did you also change the fetch in the JS to include credentials and use `no-cors`?

```js
fetch(input, {
    credentials: 'include',
    mode: 'no-cors',
})
```

**Comment by gbj** [comment]
Do you mean modifying the JS that's generated by wasm-bindgen? I'm not sure there's a reasonable way to do that at a framework level, although it's certainly possible for you as a user to patch the JS in that way after each build.

**Comment by 617a7a** [comment]
Yes, I think that would be the only way to get around this. You'd also have to disable the cross origin in Leptos for this to work though. Might open an issue with wasm-bindgen.

**Comment by gbj** [comment]
I'm going to close this as an upstream wasm-bindgen/Safari issue.

**Comment by BrandonDyer64** [comment]
Perhaps, (until there's a proper fix) the `<link rel="preload" href="/pkg/site.wasm" as="fetch" type="application/wasm" crossorigin="">` line could be excluded from the HTML response during SSR if Safari is detected in the user agent.

Is that something that could be done by leptos users, or does that need to be upstream?

**Comment by benwis** [comment]
Leptos 0.7 includes this ability. With 0.6 you’d need to modify the common leptos integration.

**Comment by prsutherland** [comment]
It looks like this is the safari/webkit issue: https://bugs.webkit.org/show_bug.cgi?id=236009

**Comment by PeteX** [comment]
I've found a workaround for this. I've removed the `<link rel="preload" ...>` headers, and changed the Javascript snippet as follows:

```
<script>
  const wasm = fetch("/pkg/mysite.wasm");
  import("/pkg/mysite.js").then(async (mod) => {
    if (document.readyState === "loading")
      await new Promise(resolve => document.addEventListener("DOMContentLoaded", resolve));

    await mod.default({ module_or_path: wasm });
    mod.hydrate();
  });
</script>
```

The snippet goes in the place where the preload headers used to be, so it gets seen by the browser at the same time. As a non-module script, it runs immediately. `fetch()` kicks off the wasm download, and it's non-blocking because it isn't awaited. The block beginning with `import` begins downloading the Javascript. It is also a promise which is not awaited, so it will not block and will run to completion in its own time.

I've done some experiments in Chrome developer tools where I've made the latency artificially bad, to highlight what is loading in each round trip. As expected, the WASM and Javascript files load together, immediately after the HTML page.
